<!-- -->
<launch>

	<arg name="mapSource" 					default="0" /> <!-- Autoware=0, Vector Map Folder=1, kml=2, lanelet2 file=3, kml map file path from runtime manager = 4, OpenDRIVE file = 6 -->
	<arg name="mapFileName" 				default="" />
	<arg name="lanelet2_origin"				default="0,0,0," /> <!-- GPS origin for lanelet2 map files, Lat, Long, Alt  -->
		
	<arg name="pathDensity" 				default="0.5" />
	<arg name="rollOutDensity" 				default="0.5" />
	<arg name="rollOutsNumber" 				default="6" />
		
	<arg name="maxLocalPlanDistance" 		default="50" />	
	<arg name="horizonDistance" 			default="200" />
	
	<arg name="minFollowingDistance" 		default="35.0" /> <!-- should be bigger than Distance to follow -->	
	<arg name="minDistanceToAvoid" 			default="20.0" /> <!-- should be smaller than minFollowingDistance and larger than maxDistanceToAvoid -->
	<arg name="maxDistanceToAvoid" 			default="5.0" /> <!-- should be smaller than minDistanceToAvoid -->
	<arg name="speedProfileFactor"			default="1.0" /> <!-- how fast the vehicle should drive through the corners and curves, less than (1) means slower speeds in the curves, larger than (1) means faster in the curves -->   
	<arg name="curveSlowDownRatio" 			default="1.0" /> <!-- how early or late to brake for upcoming turn. 1 is automatically as openplanner calculates it, Smaller than 1 means late braking, greater than 1 means brake early, this value is a ratio -->
	
	<arg name="smoothingDataWeight"			default="0.49" />
	<arg name="smoothingSmoothWeight"		default="0.2" />	
	
	<arg name="horizontalSafetyDistance"	default="1.2" />
	<arg name="verticalSafetyDistance"		default="0.8" />
	
	<arg name="enableSwerving" 				default="true" />
	<arg name="enableFollowing" 			default="true" />	
	<arg name="enableTrafficLightBehavior" 	default="false" />
	<arg name="enableStopSignBehavior" 		default="false" />	
	<arg name="enableLaneChange" 			default="false" />	
	
	<!-- Vehicle Info, shared with controller and simulation -->
	<arg name="height" 						default="1.47"  />
	<arg name="front_length"				default="1.0"  />
	<arg name="back_length"					default="0.6"  />
	<arg name="width" 						default="2.00"  />
	<arg name="length" 						default="4.54"  />	
	<arg name="wheelBaseLength" 			default="2.7"  />
	<arg name="turningRadius"				default="5.2"  />
	<arg name="maxWheelAngle" 				default="0.42" /> <!-- radians -->
	<arg name="steeringDelay" 				default="0.3" /> <!-- time steering wheel take to turn from zero to max left or right -->
	<arg name="minPursuiteDistance" 		default="3.0" /> <!-- min distance to forward following target point on the trajectory  -->
	<arg name="maxAcceleration" 			default="3.0" /> <!-- m/s*s -->
	<arg name="maxDeceleration" 			default="-3.0" /> <!-- m/s*s -->
	<arg name="maxVelocity" 				default="10.0" /> <!-- m/s -->
	<arg name="minVelocity" 				default="0.1" /> <!-- m/s -->
	<!-- ................................................... -->
	
	<!-- Internal ACC parameters, only used inside the LocalPlanner (Behavior Selector node) -->
	<arg name="use_internal_acc" 			default="true" /> <!-- If enabled, the velocity associated with the final trajectory is the final speed profile needed to be followed by the controller, including following and stopping --> 
	<arg name="accelerationPushRatio" 		default="1.0" /> <!-- push the velocity to increase even when the velocity feedback signal from sensing is very small or delayed, contribute to final acceleration. where 1 predefined acceleration value is used --> 
	<arg name="brakingPushRatio" 			default="1.0" /> <!-- push the velocity to decrease even when the velocity feedback signal from sensing is very small or delayed, contribute to final deceleration. where 1 predefined deceleration value is used-->	
	<!-- ................................................... -->
	
	<arg name="additionalBrakingDistance" 	default="10.0" />
	<arg name="goalDiscoveryDistance" 		default="1.0" />
	
	<arg name="giveUpDistance" 				default="-1.0" />
			
	<arg name="velocitySource"				default="1" /> <!-- read velocities from (0- Odometry, 1- autoware current_velocities, 2- car_info, 3- vehicle status ) "" -->
	<arg name="vel_odom_topic"				default="/odometry" />
	<arg name="vel_curr_topic"				default="/current_velocity" />
	<arg name="vel_can_topic"				default="/can_info" />
	<arg name="vehicle_status_topic"		default="/vehicle_status" />
			
	<arg name="experimentName" 				default="" />		
	
	<arg name="objects_input_topic" 		default="/detection/contour_tracker/objects" />

	<node pkg="op_local_planner" type="op_common_params" name="op_common_params" output="screen">	
		<param name="mapSource" 				value="$(arg mapSource)" /> <!-- Autoware=0, Vector Map Folder=1, kml=2 -->
		<param name="mapFileName" 				value="$(arg mapFileName)" />
		<param name="lanelet2_origin" 			value="$(arg lanelet2_origin)" />
				
		<param name="pathDensity" 			    value="$(arg pathDensity)" />
		<param name="rollOutDensity" 			value="$(arg rollOutDensity)" />
		<param name="rollOutsNumber" 			value="$(arg rollOutsNumber)"    />
					
		<param name="maxLocalPlanDistance" 		value="$(arg maxLocalPlanDistance)" />	
		<param name="horizonDistance" 			value="$(arg horizonDistance)" />	
		<param name="minFollowingDistance" 		value="$(arg minFollowingDistance)"  /> 	
		<param name="minDistanceToAvoid" 		value="$(arg minDistanceToAvoid)" /> 
		<param name="maxDistanceToAvoid" 		value="$(arg maxDistanceToAvoid)"  /> 
		<param name="speedProfileFactor"		value="$(arg speedProfileFactor)"  />	
		<param name="smoothingDataWeight"		value="$(arg smoothingDataWeight)"  />
		<param name="smoothingSmoothWeight"		value="$(arg smoothingSmoothWeight)"  />	
		<param name="horizontalSafetyDistance"	value="$(arg horizontalSafetyDistance)"  />
		<param name="verticalSafetyDistance"	value="$(arg verticalSafetyDistance)"  />	
		<param name="enableSwerving" 			value="$(arg enableSwerving)"  />
		<param name="enableFollowing" 			value="$(arg enableFollowing)" />	
		<param name="enableTrafficLightBehavior" value="$(arg enableTrafficLightBehavior)" />
		<param name="enableStopSignBehavior" 	value="$(arg enableStopSignBehavior)" />	
		<param name="enableLaneChange" 			value="$(arg enableLaneChange)" />
		
		<param name="height" 					value="$(arg height)"  />		
		<param name="front_length" 				value="$(arg front_length)"  />
		<param name="back_length" 				value="$(arg back_length)"  />
		<param name="width" 					value="$(arg width)"  />
		<param name="length" 					value="$(arg length)"  />		
		<param name="wheelBaseLength" 			value="$(arg wheelBaseLength)"  />
		<param name="turningRadius"				value="$(arg turningRadius)"  />
		<param name="maxWheelAngle" 			value="$(arg maxWheelAngle)" />
		<param name="steeringDelay" 			value="$(arg steeringDelay)" />
		<param name="minPursuiteDistance" 		value="$(arg minPursuiteDistance)"  />
		<param name="maxAcceleration" 			value="$(arg maxAcceleration)" />
		<param name="maxDeceleration" 			value="$(arg maxDeceleration)" />
		<param name="maxVelocity" 				value="$(arg maxVelocity)" />
		<param name="minVelocity" 				value="$(arg minVelocity)" />
		<param name="use_internal_acc" 			value="$(arg use_internal_acc)" />
		<param name="accelerationPushRatio" 	value="$(arg accelerationPushRatio)" />
		<param name="brakingPushRatio" 			value="$(arg brakingPushRatio)" />
		<param name="curveSlowDownRatio" 		value="$(arg curveSlowDownRatio)" />
			
		<param name="additionalBrakingDistance" value="$(arg additionalBrakingDistance)"  />
		<param name="goalDiscoveryDistance" 	value="$(arg goalDiscoveryDistance)"  />
		<param name="giveUpDistance" 			value="$(arg giveUpDistance)"  />
		<param name="velocitySource"			value="$(arg velocitySource)" /> 
		<param name="vel_odom_topic"			value="$(arg vel_odom_topic)" />
		<param name="vel_curr_topic"			value="$(arg vel_curr_topic)" />
		<param name="vel_can_topic"				value="$(arg vel_can_topic)" />			
		<param name="vehicle_status_topic"		value="$(arg vehicle_status_topic)" />
		<param name="experimentName" 			value="$(arg experimentName)" />	
		<param name="objects_input_topic" 		value="$(arg objects_input_topic)" />	
	</node>				
			
</launch>
